### PR TITLE
Add `ChooseParentView` to `PageListingViewSet`

### PR DESCRIPTION
**Pull request by abdlrhman08**
Added `ChooseParentView` for #11734. The optimzation for `get_valid_parents()` is still needed. Any ideas on how to optimize `get_valid_parents()` would be really helpful !

### CODE REVIEW

**Comment by abdlrhman08** [comment]
The backend tests are failing because the exception message raised in `copy_action` is not the same as the one asserted in `wagtail.admin.tests.api.test_pages`

**Comment by abdlrhman08** [comment]
All tests now pass successfully

File: wagtail/admin/viewsets/pages.py [hunk 1]
```diff
@@ -29,12 +32,19 @@ def get_index_view_kwargs(self, **kwargs):
...
             "index_url_name": self.get_url_name("index"),
             "index_results_url_name": self.get_url_name("index_results"),
+            "choose_parent_url_name": self.get_url_name("choose_parent"),
             "model": self.model,
             "columns": self.columns,
             "filterset_class": self.filterset_class,
             **kwargs,
         }
 
+    def get_chooseparent_view_kwargs(self, **kwargs):
```
Comment (laymonage) [inline wagtail/admin/viewsets/pages.py:42]:
```suggestion
    def get_choose_parent_view_kwargs(self, **kwargs):
```

File: wagtail/admin/viewsets/pages.py [hunk 2]
```diff
@@ -47,8 +57,15 @@ def index_results_view(self):
             self.index_view_class, **self.get_index_view_kwargs(), results_only=True
         )
 
+    @property
+    def choose_parent_view(self):
+        return self.construct_view(
+            self.choose_parent_view_class, **self.get_chooseparent_view_kwargs()
```
Comment (laymonage) [inline wagtail/admin/viewsets/pages.py:63]:
```suggestion
            self.choose_parent_view_class, **self.get_choose_parent_view_kwargs()
```

File: wagtail/admin/templates/wagtailadmin/shared/headers/page_viewset_header.html [hunk 3]
```diff
@@ -0,0 +1,7 @@
+{% extends 'wagtailadmin/shared/headers/slim_header.html' %}
+{% load wagtailadmin_tags i18n %}
+
+{% block actions %}
+    {% trans 'Create new page' as create_page_str %}
+    <a href="{{ choose_parent_url }}" class="w-header-button" data-controller="w-tooltip" data-w-tooltip-content-value="{{ create_page_str }}" aria-label="{{ create_page_str }}">{% icon name="plus" %}</a>
+{% endblock %}
```
Comment (laymonage) [inline wagtail/admin/templates/wagtailadmin/shared/headers/page_viewset_header.html:7]: We should avoid making new header templates when necessary. It's possible to add custom buttons from the view code using the `header_buttons` property.

File: wagtail/admin/templates/wagtailadmin/pages/index.html [hunk 4]
```diff
@@ -2,7 +2,7 @@
 {% load wagtailadmin_tags i18n %}
 
 {% block header %}
-    {% include 'wagtailadmin/shared/headers/slim_header.html' with title=header_title search_url=index_results_url %}
+    {% include 'wagtailadmin/shared/headers/page_viewset_header.html' with title=header_title search_url=index_results_url %}
```
Comment (laymonage) [inline wagtail/admin/templates/wagtailadmin/pages/index.html:5]: Let's try to keep using `slim_header.html` if possible 
```suggestion
    {% include 'wagtailadmin/shared/headers/slim_header.html' with title=header_title search_url=index_results_url %}
```

File: wagtail/admin/templates/wagtailadmin/pages/choose_parent.html [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+{% block titletag %}Choose a parent page{% endblock %}
+
+{% block extra_css %}
+    {{ block.super }}
+    {{ form.media.css }}
+{% endblock %}
+
+{% block extra_js %}
+    {{ block.super }}
+    {% include "wagtailadmin/pages/_editor_js.html" %}
```
Comment (laymonage) [inline wagtail/admin/templates/wagtailadmin/pages/choose_parent.html:13]:
```suggestion
```

This is probably no longer necessary now that we include the modal-workflow.js for the chooser is included everywhere. Could you recheck?

File: wagtail/admin/templates/wagtailadmin/pages/choose_parent.html [hunk 6]
```diff
@@ -0,0 +1,31 @@
+{% extends "wagtailadmin/base.html" %}
+{% load i18n wagtailadmin_tags%}
```
Comment (laymonage) [inline wagtail/admin/templates/wagtailadmin/pages/choose_parent.html:2]:
```suggestion
{% load i18n wagtailadmin_tags %}
```

File: wagtail/admin/templates/wagtailadmin/pages/choose_parent.html [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    {% block header %}
+        {% include "wagtailadmin/shared/header.html" with title=_("Create a new ") subtitle=model_name icon="doc-empty-inverse" %}
+    {% endblock %}
+
+    <div class="nice-padding">
+        <h2>{% blocktrans trimmed %}Choose a parent page{% endblocktrans %}</h2>
+        <form action="" method="post" novalidate>
+            {% csrf_token %}
+
+            {% include "wagtailadmin/shared/field.html" with field=form.parent_page %}
```
Comment (laymonage) [inline wagtail/admin/templates/wagtailadmin/pages/choose_parent.html:27]: To allow people customising the form more easily

```suggestion
            {% for field in form %}
                {% formattedfield field %}
            {% endfor %}
```

File: wagtail/admin/templates/wagtailadmin/pages/choose_parent.html [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+{% block extra_js %}
+    {{ block.super }}
+    {% include "wagtailadmin/pages/_editor_js.html" %}
+    {{ form.media.js }}
+{% endblock %}
+
+{% block content %}
+    {% block header %}
+        {% include "wagtailadmin/shared/header.html" with title=_("Create a new ") subtitle=model_name icon="doc-empty-inverse" %}
+    {% endblock %}
```
Comment (laymonage) [inline wagtail/admin/templates/wagtailadmin/pages/choose_parent.html:20]: I can't explain the details right now, but ideally we want to move away from `header.html` in favour of `slim_header.html` with the breadcrumbs. See how we render the `wagtail.admin.views.generic.models.CreateView` and its corresponding `wagtailadmin/generic/form.html` template for example.

```suggestion
{% block main_content %}
```

File: wagtail/admin/views/pages/choose_parent.py [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+            content_type__in=allowed_parent_page_content_types
+        )
+
+        # Get queryset of pages where the user has permission to add subpages
+        if user.is_superuser:
+            pages_where_user_can_add = Page.objects.all()
+        else:
+            pages_where_user_can_add = Page.objects.none()
+
+            from wagtail.permission_policies.pages import PagePermissionPolicy
```
Comment (laymonage) [inline wagtail/admin/views/pages/choose_parent.py:38]: We should use the singular instance of the policy as of #11267

```suggestion
            from wagtail.permissions import page_permission_policy
```

and move this to the top-level imports

File: wagtail/admin/views/pages/choose_parent.py [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+            perms = {
+                perm
+                for perm in PagePermissionPolicy().get_cached_permissions_for_user(user)
```
Comment (laymonage) [inline wagtail/admin/views/pages/choose_parent.py:42]:
```suggestion
                for perm in page_permission_policy.get_cached_permissions_for_user(user)
```

File: wagtail/admin/viewsets/pages.py [hunk 11]
```diff
@@ -29,12 +32,19 @@ def get_index_view_kwargs(self, **kwargs):
         return {
             "index_url_name": self.get_url_name("index"),
             "index_results_url_name": self.get_url_name("index_results"),
+            "choose_parent_url_name": self.get_url_name("choose_parent"),
```
Comment (laymonage) [inline wagtail/admin/viewsets/pages.py:35]: Have you checked whether it's possible to just supply this as `add_url_name`, and see whether the view automatically picks it up and renders the header button accordingly?

```suggestion
            "add_url_name": self.get_url_name("choose_parent"),
```

If so, we can remove `choose_parent_url_name` and `get_choose_parent_url` from the listing view.

**Review by laymonage — COMMENTED** [review]
Thanks, this is a really good start! I haven't tested this but I think it's worth making sure that we use the established patterns for new views in Wagtail instead of following the ModelAdmin implementation as-is. I've left comments on how to achieve some of these things.

Don't worry about the `get_valid_parent_pages` performance optimisation just yet, we can look into it once we're happy with the rest of the view's implementation.

File: wagtail/admin/forms/pages.py [hunk 12]
```diff
@@ -226,3 +226,31 @@ def __init__(self, *args, **kwargs):
             label=_("New parent page"),
             help_text=_("Select a new parent for this page."),
         )
+
+
+class PageChoiceField(forms.ModelChoiceField):
+    def label_from_instance(self, obj):
+        if obj.is_root():
+            return obj.get_admin_display_title()
```
Comment (abdlrhman08) [inline wagtail/admin/forms/pages.py:234]: We exclude the root page when we are trying to get the ancestor, this leads to the label for the root page itself rendering as blank, so I had to add this.

Or should we prevent the user from choosing the user from choosing the root here as a parent ?

**Comment by abdlrhman08** [comment]
Thanks for your review @laymonage ! I made the changes and reverted the header back to `slim_header.html`, and also added breadcrumb items. Once we're happy with the view's implementation, I'll immediately start writing the tests. I left a comment if you can check it.

I can see you self-assigned this PR to your self, I am not sure if you are going to continue with working on this, but I'd be really happy to help or to do anything I can do :)

**Review by laymonage — COMMENTED** [review]
Thanks @abdlrhman08!

There's a few more improvements we can make around the view code and the template, so they align better with Django's and our generic views.

Also, it seems we can replace the choice field with a page chooser, which makes it very much in line with our existing solutions e.g. for copying a page and for moving a page.

I've made these adjustments locally, so I'll push them to your branch. I haven't written the tests, though. Would you be up for writing them? Thanks again!

**Comment by abdlrhman08** [comment]
For some reason I didn't get notified about your changes, sorry for seeing them late. I think this is going great. I'd be happy to write the tests. Thank you for your help on my PRs @laymonage !

**Comment by abdlrhman08** [comment]
I added two tests, one for the chooser view itself which gets asserted with the template, and another test to check for correct redirects. Let me know if any other tests are needed. If any other test case comes to mind I'll write it.

**Comment by abdlrhman08** [comment]
Added two more test cases and created a separate class for them

File: wagtail/admin/tests/pages/test_parent_page_chooser_view.py [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+    def setUp(self):
+        super().setUp()
+        self.login()
+
+        self.view_url = reverse("event_pages:choose_parent")
+
+    def test_get_page_parent_chooser(self):
+        response = self.client.get(self.view_url)
+        self.assertEqual(response.status_code, 200)
+        self.assertTemplateUsed(response, "wagtailadmin/pages/choose_parent.html")
```
Comment (laymonage) [inline wagtail/admin/tests/pages/test_parent_page_chooser_view.py:20]:
```suggestion
        self.assertTemplateUsed(response, "wagtailadmin/pages/choose_parent.html")
        self.assertBreadcrumbsItemsRendered(
            [
                {"url": reverse("event_pages:index"), "label": "Event pages"},
                {"url": "", "label": "Choose parent", "sublabel": "Event page"},
            ],
            response.content,
        )
```

File: wagtail/admin/views/chooser.py [hunk 14]
```diff
@@ -93,7 +93,7 @@ def can_choose_page(
 
             if user_perm == "move_to":
                 return page_to_move.permissions_for_user(user).can_move_to(page)
-    if user_perm == "copy_to":
+    if user_perm in {"add_subpage", "copy_to"}:
         return page.permissions_for_user(user).can_add_subpage()
```
Comment (laymonage) [inline wagtail/admin/views/chooser.py:97]: Unfortunately this doesn't take into account the allowed `subpage_types` and things like `max_count_per_parent`. Adding these checks here is quite complicated as it involves passing the params to the JS code, to later be used by the chooser views. I think for now we can validate this via the form instead.

**Review by laymonage — COMMENTED** [review]
Thanks for giving a go at the tests! I noticed that in some edge cases, the user can select a page and click "Create", only to be redirected to the admin home page with a permission denied message. Unfortunately there's a limitation to how detailed we can make the permission checks in the choosers, and adding more logic to it is a bit more involved for this PR.

An alternative that we can do is validating the selected page via the form validation mechanism. I've left a few comments on how to do this. We'll need to write the tests for it too, though. Feel free to give it a try, but I can also take over if you'd like.

Thanks again!

**Comment by abdlrhman08** [comment]
I added the form validation and also wrote the tests for them. I used your implementation and everything now works correctly. I want to grab your attention to something, in the page chooser you can by-pass the user permissions and any page can be selected if you just change the locale selected in the chooser.

https://github.com/wagtail/wagtail/assets/69439342/532892d6-d530-441b-9053-93939c55fd46

The form will still validate the page in the post request, but I think it would be better if choosing it is prevented in any way. I fixed it in the last commit, have a look. 

> Feel free to give it a try, but I can also take over if you'd like.
I am always happy to help, but if there is some kind of a deadline for this feature or something feel free to take over.

Thank you again for your help in my PRs !

File: wagtail/admin/views/chooser.py [hunk 15]
```diff
@@ -298,22 +298,23 @@ def get(self, request, parent_page_id=None):
...
+                        "page_type": page_type_string,
+                        "locale": locale.language_code,
+                        "user_perms": user_perm,
                     }
-                    for locale in Locale.objects.filter(
-                        pk__in=pages.values_list("locale_id")
-                    ).exclude(pk=active_locale_id)
-                ]
+                    if not user_perm:
+                        del params_dict["user_perms"]
```
Comment (laymonage) [inline wagtail/admin/views/chooser.py:310]: Thanks for highlighting the issue! It seems this is not the only query parameter that gets lost during the locale switching. Instead of specifying each query param (and potentially forgetting to do this for future query params), we could copy the current request's query params and remove ones that aren't necessary. I noticed that in other places we tend to do this, e.g.:

https://github.com/wagtail/wagtail/blob/dd5e16b2e19c63c0764da8d5f2a16661386bf73f/wagtail/admin/templates/wagtailadmin/chooser/_browse_results.html#L18

https://github.com/wagtail/wagtail/blob/dd5e16b2e19c63c0764da8d5f2a16661386bf73f/wagtail/admin/templates/wagtailadmin/chooser/tables/page_navigate_to_children_cell.html#L4

File: wagtail/admin/views/pages/choose_parent.py [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+
+    @cached_property
+    def submit_button_label(self):
+        return _("Create a new %(model_name)s") % {
+            "model_name": self.model._meta.verbose_name,
+        }
+
+    def form_valid(self, form):
+        opts = self.model._meta
+
```
Comment (laymonage) [inline wagtail/admin/views/pages/choose_parent.py:61]:
```suggestion
```

File: wagtail/admin/templates/wagtailadmin/pages/index.html [hunk 17]
```diff
@@ -2,7 +2,7 @@
 {% load wagtailadmin_tags i18n %}
 
 {% block header %}
-    {% include 'wagtailadmin/shared/headers/slim_header.html' with title=header_title search_url=index_results_url %}
+    {% include 'wagtailadmin/shared/headers/slim_header.html' with title=header_title search_url=index_results_url buttons=header_buttons %}
```
Comment (laymonage) [inline wagtail/admin/templates/wagtailadmin/pages/index.html:5]: While we're here, let's add icon support for the listing view as well

```suggestion
    {% include 'wagtailadmin/shared/headers/slim_header.html' with title=header_title icon_name=header_icon search_url=index_results_url buttons=header_buttons %}
```

**Review by laymonage — COMMENTED** [review]
This is looking really good! I'm proposing a different way to fix the locale selector issue, but otherwise this looks good to me. I've made the edits locally so I'll push them and merge. Thanks!

**Review by laymonage — APPROVED** [review]
